<a href="https://colab.research.google.com/github/byfurkation/TelecomX_datos/blob/main/TelecomX_LATAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Telecom X - Análisis de Evasión de Clientes

Has sido contratado como asistente de análisis de datos en Telecom X y formarás parte del proyecto "Churn de Clientes". **La empresa enfrenta una alta tasa de cancelaciones y necesita comprender los factores que llevan a la pérdida de clientes.**

Tu desafío será recopilar, procesar y analizar los datos, utilizando Python y sus principales bibliotecas para extraer información valiosa. A partir de tu análisis, el equipo de Data Science podrá avanzar en modelos predictivos y desarrollar estrategias para reducir la evasión.

¿Qué vas a practicar?

✅ Importar y manipular datos desde una API de manera eficiente.

✅ Aplicar los conceptos de ETL (Extracción, Transformación y Carga) en la preparación de los datos.

✅ Crear visualizaciones estratégicas para identificar patrones y tendencias.

✅ Realizar un Análisis Exploratorio de Datos (EDA) y generar un informe con insights relevantes.

¡Ahora es tu turno! 🚀 Usa tus conocimientos para transformar datos en información estratégica y ayudar a Telecom X a retener más clientes.

##📌 Extracción

Para iniciar tu análisis, necesitarás importar los datos de la API de Telecom X. Estos datos están disponibles en formato JSON y contienen información esencial sobre los clientes, incluyendo datos demográficos, tipo de servicio contratado y estado de evasión.

¿Qué debes hacer?

✅ Cargar los datos directamente desde la API utilizando Python.

✅ Convertir los datos a un DataFrame de Pandas para facilitar su manipulación.

Este es el primer paso para transformar los datos en información valiosa. ¿Listo para programar? 🚀

### **Ayuda a visualizar un archivo un formato Json en la web.**

https://jsonviewer.stack.hu/



In [1]:
# Procederemos a cargar un archivo Json, Por lo cual como primer paso
# necesitamos importar pandas y Json
import pandas as pd
import json

In [2]:
df = pd.read_json('https://raw.githubusercontent.com/byfurkation/TelecomX_datos/d3d77c77396a50d0951630ec521dc3b048607b6f/TelecomX_Data.json')

df.head(10)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
5  0013-MHZWF    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
6  0013-SMEOE    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
7  0014-BMAQU    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
8  0015-UOCOJ    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
9  0016-QLJIS    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
8   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
9  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   
5  {'InternetService': 'DSL', 'OnlineSecurity': '...   
6  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7  {'InternetService': 'Fiber optic', 'OnlineSecu...   
8  {'InternetService': 'DSL', 'OnlineSecurity': '...   
9  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...  
5  {'Contract': 'Month-to-month', 'PaperlessBilli...  
6  {'Contract': 'Two year', 'PaperlessBilling': '...  
7  {'Contract': 'Two year', 'PaperlessBilling': '...  
8  {'Contract': 'Month-to-month', 'PaperlessBilli...  
9  {'Contract': 'Two year', 'PaperlessBilling': '...

### **Resultado**
De esta manera queda guardado automáticamente en un Data frame nuestra información, pues al aplicarle el método read a cualquier importación de archivos se transforma automáticamente en uno.

##🔧 Transformación

In [3]:
# Normalizamos cada columna con diccionarios
df_customer = pd.json_normalize(df['customer'])
df_phone    = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account  = pd.json_normalize(df['account'])

# Concatenamos todas las columnas nuevas con el DataFrame original
df_normalizado = pd.concat([df.drop(['customer', 'phone', 'internet', 'account'], axis=1),
                      df_customer, df_phone, df_internet, df_account], axis=1)


In [4]:
df_normalizado.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

In [5]:
df_normalizado.sample(20)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
1165  1658-XUHBX    No  Female              1     Yes        Yes      59   
1364  1972-RSMOV        Female              0     Yes         No      40   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   
5284  7218-HKQFK   Yes    Male              0     Yes         No       2   
567   0812-WUPTB   Yes    Male              1     Yes         No       1   
6631  9114-DPSIA    No    Male              0     Yes        Yes      72   
4500  6169-PGNCD    No  Female              0      No         No      57   
1232  1763-KUAAW    No  Female              1      No         No      18   
3838  5248-RPYWW    No  Female              1     Yes        Yes      72   
1268  1821-BUCWY    No    Male              0      No         No      30   
617   0876-WDUUZ   Yes  Female              0      No         No       5   
3958  5404-GGUKR    No    Male              0      No         No       5   
1884  2672-TGEFF    No  Female              0     Yes        Yes      37   
2141  2996-XAUVF    No    Male              0      No         No      70   
2814  3900-AQPHZ   Yes  Female              0      No         No       1   
2322  3235-ETOOB   Yes    Male              0     Yes         No       8   
4882  6682-QJDGB   Yes    Male              0      No        Yes      40   
6429  8823-RLPWL    No  Female              0     Yes         No      72   
6941  9564-KCLHR   Yes    Male              0      No         No       1   
5091  6961-MJKBO    No    Male              0      No         No      47   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
1165          Yes               Yes     Fiber optic  ...                   No   
1364          Yes                No              No  ...  No internet service   
5599          Yes                No              No  ...  No internet service   
5284          Yes                No     Fiber optic  ...                   No   
567           Yes                No     Fiber optic  ...                   No   
6631          Yes               Yes             DSL  ...                  Yes   
4500          Yes                No             DSL  ...                  Yes   
1232          Yes                No              No  ...  No internet service   
3838          Yes               Yes             DSL  ...                  Yes   
1268          Yes                No             DSL  ...                   No   
617           Yes                No     Fiber optic  ...                   No   
3958          Yes                No             DSL  ...                  Yes   
1884          Yes               Yes     Fiber optic  ...                  Yes   
2141           No  No phone service             DSL  ...                   No   
2814          Yes                No              No  ...  No internet service   
2322          Yes                No     Fiber optic  ...                   No   
4882          Yes                No             DSL  ...                  Yes   
6429          Yes               Yes     Fiber optic  ...                  Yes   
6941          Yes                No             DSL  ...                   No   
5091          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
1165                  Yes                   No                   No   
1364  No internet service  No internet service  No internet service   
5599  No internet service  No internet service  No internet service   
5284                   No                  Yes                  Yes   
567                    No                   No                   No   
6631                  Yes                  Yes                   No   
4500                  Yes                  Yes                  Yes   
1232  No internet service  No internet service  No internet service   
3838                  Yes                  Yes                  Yes 

### **Conoce el conjunto de datos**

Ahora que has extraído los datos, es fundamental comprender la estructura del dataset y el significado de sus columnas. Esta etapa te ayudará a identificar qué variables son más relevantes para el análisis de evasión de clientes.

📌 Para facilitar este proceso, se creó un diccionario de datos con la descripción de cada columna, mismo que se encuentra en el repositorio bajo el nombre de "TelecomX_diccionario.md".



### **Diccionario de datos**

customerID: número de identificación único de cada cliente

Churn: si el cliente dejó o no la empresa

gender: género (masculino y femenino)

SeniorCitizen: información sobre si un cliente tiene o no una edad igual o mayor a 65 años

Partner: si el cliente tiene o no una pareja

Dependents: si el cliente tiene o no dependientes

tenure: meses de contrato del cliente

PhoneService: suscripción al servicio telefónico

MultipleLines: suscripción a más de una línea telefónica

InternetService: suscripción a un proveedor de internet

OnlineSecurity: suscripción adicional de seguridad en línea

OnlineBackup: suscripción adicional de respaldo en línea

DeviceProtection: suscripción adicional de protección del dispositivo

TechSupport: suscripción adicional de soporte técnico, menor tiempo de espera

StreamingTV: suscripción de televisión por cable

StreamingMovies: suscripción de streaming de películas

Contract: tipo de contrato

PaperlessBilling: si el cliente prefiere recibir la factura en línea

PaymentMethod: forma de pago

Charges.Monthly: total de todos los servicios del cliente por mes

Charges.Total: total gastado por el cliente

### **Tarea 1:** Explorar las columnas del dataset y verificar sus tipos de datos.


In [6]:
# Observamos que la totalidad de su registro son 7,267, y en sus columnas cada
# una registra los mismos con los datos no nulo.
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


### **Tarea 2**: Consultar el diccionario para comprender mejor el significado de las variables.


In [7]:
# Valores unicos de una columna
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [8]:
# Ver valores únicos y sus frecuencias para todas las columnas

for col in df_normalizado.columns:
    print(f"--- {col} ---")
    print(df_normalizado[col].value_counts(dropna=False))  # incluye NaN si existen
    print()

--- customerID ---
customerID
9995-HOTOH    1
0002-ORFBO    1
0003-MKNFE    1
9970-QBCDA    1
9968-FFVVH    1
             ..
0014-BMAQU    1
0013-SMEOE    1
0013-MHZWF    1
0013-EXCHZ    1
0011-IGKFF    1
Name: count, Length: 7267, dtype: int64

--- Churn ---
Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

--- gender ---
gender
Male      3675
Female    3592
Name: count, dtype: int64

--- SeniorCitizen ---
SeniorCitizen
0    6085
1    1182
Name: count, dtype: int64

--- Partner ---
Partner
No     3749
Yes    3518
Name: count, dtype: int64

--- Dependents ---
Dependents
No     5086
Yes    2181
Name: count, dtype: int64

--- tenure ---
tenure
1     634
72    369
2     246
3     207
4     185
     ... 
28     60
39     59
44     54
36     50
0      11
Name: count, Length: 73, dtype: int64

--- PhoneService ---
PhoneService
Yes    6560
No      707
Name: count, dtype: int64

--- MultipleLines ---
MultipleLines
No                  3495
Yes                 3065
No phone s

#### **Decisión de análisis clave**

Antes de transformar o eliminar datos, se debe entender qué significan los valores faltantes y cómo afectan el análisis. Como en este caso es un ejercicio didáctico, podemos aplicar criterios lógicos con justificación.

**Columna: Churn**

Valores únicos:

"No": 5174

"Yes": 1869

" " (espacios vacíos): 224

Esta **columna define si un cliente abandonó o no el servicio**, lo cual es clave para análisis predictivo.

Se eliminaránn 224 registros (~3%) con datos faltantes en la columna Churn, ya que esta representa la variable objetivo del análisis y no se puede inferir razonablemente su valor.

In [9]:
# eliminar filas vacías de churn 224+7043= 7, 267

# Eliminar filas donde Churn es solo un espacio
df_normalizado = df_normalizado[df_normalizado['Churn'].str.strip() != '']
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7043 non-null   object 
 3   SeniorCitizen     7043 non-null   int64  
 4   Partner           7043 non-null   object 
 5   Dependents        7043 non-null   object 
 6   tenure            7043 non-null   int64  
 7   PhoneService      7043 non-null   object 
 8   MultipleLines     7043 non-null   object 
 9   InternetService   7043 non-null   object 
 10  OnlineSecurity    7043 non-null   object 
 11  OnlineBackup      7043 non-null   object 
 12  DeviceProtection  7043 non-null   object 
 13  TechSupport       7043 non-null   object 
 14  StreamingTV       7043 non-null   object 
 15  StreamingMovies   7043 non-null   object 
 16  Contract          7043 non-null   object 
 17  

### **Tarea 3:** Identificar las columnas más relevantes para el análisis de evasión.

🎯 Objetivo asumido:

Analizar y predecir el "Churn" (si un cliente se va o se queda)

1. tenure (meses de contrato)

🧠 Relevancia: Refleja el tiempo de permanencia. Clientes con menor tenure son más propensos a irse.

🧪 Análisis: Puedes hacer histogramas, segmentaciones, correlaciones con Churn.

2. Contract (tipo de contrato: mensual, anual...)

🧠 Relevancia: Clientes con contrato mes a mes tienden a irse más fácilmente.

📊 Ideal para gráficos de barras o análisis de riesgo.

3. PaymentMethod (forma de pago)

💳 Pagos automáticos vs. manuales podrían influir.

🎯 A menudo se asocia a abandono por fricción en pagos.

4. Charges.Monthly y Charges.Total

💵 Clientes con cargos más altos pueden estar menos satisfechos.

También pueden indicar uso intensivo (más fidelidad).

Útiles para crear grupos de riesgo por gasto.

5. InternetService y servicios relacionados:

InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies

🧠 Estos servicios son opcionales y reflejan nivel de compromiso y satisfacción.

Algunos modelos han demostrado que la combinación de servicios influye fuertemente en la retención.

6. PaperlessBilling

📥 Puede parecer menor, pero se ha encontrado correlación entre este tipo de facturación y cancelación: refleja preferencia por lo digital y menos fricción al cambiar de proveedor.

7. PhoneService y MultipleLines

🔁 Relacionados con fidelización: clientes que contratan más de un servicio tienden a quedarse más.

⚖️ Variables demográficas (prioridad media)

Estas no siempre son predictoras fuertes, pero pueden dar contexto:

SeniorCitizen: Algunos estudios muestran que personas mayores son más estables en sus servicios.

Partner y Dependents: Pueden reflejar estabilidad o necesidad de varios servicios.

gender: Usualmente no es significativa para Churn por sí sola, pero puede considerarse por completitud.




### **Tarea 4** Comprobación de incoherencias en los datos

En este paso, verifica si hay problemas en los datos que puedan afectar el análisis. Presta atención a valores ausentes, duplicados, errores de formato e inconsistencias en las categorías. Este proceso es esencial para asegurarte de que los datos estén listos para las siguientes etapas.

 0   customerID        7043 non-null   object ok

 1   Churn             7043 non-null   object boolean

 2   gender            7043 non-null   object ok

 3   SeniorCitizen     7043 non-null   int64  ok

 4   Partner           7043 non-null   object boolean

 5   Dependents        7043 non-null   object boolean

 6   tenure            7043 non-null   int64  ok

 7   PhoneService      7043 non-null   object int64

 8   MultipleLines     7043 non-null   object boolean

 9   InternetService   7043 non-null   object boolean

 10  OnlineSecurity    7043 non-null   object boolean

 11  OnlineBackup      7043 non-null   object boolean

 12  DeviceProtection  7043 non-null   object boolean

 13  TechSupport       7043 non-null   object boolean

 14  StreamingTV       7043 non-null   object boolean

 15  StreamingMovies   7043 non-null   object boolean

 16  Contract          7043 non-null   object ok

 17  PaperlessBilling  7043 non-null   object boolean

 18  PaymentMethod     7043 non-null   object ok

 19  Charges.Monthly   7043 non-null   float64 ok

 20  Charges.Total     7043 non-null   object float 64

### **Tarea 5** Manejo de inconsistencias

Ahora que has identificado las inconsistencias, es momento de aplicar las correcciones necesarias. Ajusta los datos para asegurarte de que estén completos y coherentes, preparándolos para las siguientes etapas del análisis.

In [10]:
# El método .copy() solo se ejecuta una vez para cortar cualquier vínculo con
# una vista temporal y evitar las advertencias como SettingWithCopyWarning.
df_normalizado = df_normalizado.copy()

In [11]:
# limpiar valores no numéricos
df_normalizado['Charges.Total'] = df_normalizado['Charges.Total'].replace(' ', pd.NA)

In [12]:
df_normalizado['Charges.Total'] = pd.to_numeric(df_normalizado['Charges.Total'], errors='coerce')

In [13]:
df_normalizado['Charges.Total'].dtypes  # Debería devolver: dtype('float64')

dtype('float64')

In [14]:
# 1. Churn: "Yes" → True, "No" → False
df_normalizado['Churn'] = df_normalizado['Churn'].map({'Yes': True, 'No': False})

# 2. SeniorCitizen: 1 → True, 0 → False
df_normalizado['SeniorCitizen'] = df_normalizado['SeniorCitizen'].map({1: True, 0: False})

# 3. Partner: "Yes" → True, "No" → False
df_normalizado['Partner'] = df_normalizado['Partner'].map({'Yes': True, 'No': False})

# 4. Dependents: "Yes" → True, "No" → False
df_normalizado['Dependents'] = df_normalizado['Dependents'].map({'Yes': True, 'No': False})

# 5. PhoneService: "Yes" → True, "No" → False
df_normalizado['PhoneService'] = df_normalizado['PhoneService'].map({'Yes': True, 'No': False})

# 6. PaperlessBilling: "Yes" → True, "No" → False
df_normalizado['PaperlessBilling'] = df_normalizado['PaperlessBilling'].map({'Yes': True, 'No': False})

In [15]:
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   bool   
 2   gender            7043 non-null   object 
 3   SeniorCitizen     7043 non-null   bool   
 4   Partner           7043 non-null   bool   
 5   Dependents        7043 non-null   bool   
 6   tenure            7043 non-null   int64  
 7   PhoneService      7043 non-null   bool   
 8   MultipleLines     7043 non-null   object 
 9   InternetService   7043 non-null   object 
 10  OnlineSecurity    7043 non-null   object 
 11  OnlineBackup      7043 non-null   object 
 12  DeviceProtection  7043 non-null   object 
 13  TechSupport       7043 non-null   object 
 14  StreamingTV       7043 non-null   object 
 15  StreamingMovies   7043 non-null   object 
 16  Contract          7043 non-null   object 
 17  

In [16]:
# La diferencia de 11 valores (7043 vs. 7032) confirma que 11 celdas estaban
# vacías o corruptas.
df_normalizado[df_normalizado['Charges.Total'].isna()]

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
975   1371-DWPAZ  False  Female          False     True        True       0   
1775  2520-SGTTA  False  Female          False     True        True       0   
1955  2775-SEFEE  False    Male          False    False        True       0   
2075  2923-ARZLG  False    Male          False     True        True       0   
2232  3115-CZMZD  False    Male          False    False        True       0   
2308  3213-VVOLG  False    Male          False     True        True       0   
2930  4075-WKNIU  False  Female          False     True        True       0   
3134  4367-NUYAO  False    Male          False     True        True       0   
3203  4472-LVYGI  False  Female          False     True        True       0   
4169  5709-LVOEQ  False  Female          False     True        True       0   
5599  7644-OMVMY  False    Male          False     True        True       0   

      PhoneService     MultipleLines InternetService  ...  \
975          False  No phone service             DSL  ...   
1775          True                No              No  ...   
1955          True               Yes             DSL  ...   
2075          True                No              No  ...   
2232          True                No              No  ...   
2308          True               Yes              No  ...   
2930          True               Yes             DSL  ...   
3134          True               Yes              No  ...   
3203         False  No phone service             DSL  ...   
4169          True                No             DSL  ...   
5599          True                No              No  ...   

             OnlineBackup     DeviceProtection          TechSupport  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                  Yes                   No                  Yes   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                   No                  Yes                  Yes   
4169                  Yes                  Yes                   No   
5599  No internet service  No internet service  No internet service   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
975                   Yes                   No  Two year            False   
1775  No internet service  No internet service  Two year            False   
1955                   No                   No  Two year             True   
2075  No internet service  No internet service  One year             True   
2232  No internet service  No internet service  Two year            False   
2308  No internet service  No internet service  Two year            False   
2930                  Yes                   No  Two year            False   
3134  No internet service  No internet service  Two year            False   
3203                  Yes                   No  Two year             True   
4169                  Yes                  Yes  Two year            False   
5599  No internet service  No internet service  Two year            False   

                  PaymentMethod Charges.Monthly  Charges.Total  
975     Credit card (automatic)           56.05            NaN  
1775               Mailed check           20.00            NaN  
1955  Bank transfer (automatic)           61.90            NaN  
2075               Mailed check           19.70            NaN  
2232               Mailed check           20.25            NaN  
2308               Mailed check           25.35            NaN  
2930               Mailed check           73.35            NaN  
3134               Mailed 

In [17]:
# Razón por la cual al no ser muchas filas se optará por eliminarlas ya que no se puede obtener el valor total de ninguna manera clara. En otras circunstancias verificaríamos con los equipos de venta y verificaríamos cada una de las categorías a fin de saber cuál es el cargo de cada uno de los clientes.

df_normalizado = df_normalizado.dropna(subset=['Charges.Total'])

In [18]:
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   Churn             7032 non-null   bool   
 2   gender            7032 non-null   object 
 3   SeniorCitizen     7032 non-null   bool   
 4   Partner           7032 non-null   bool   
 5   Dependents        7032 non-null   bool   
 6   tenure            7032 non-null   int64  
 7   PhoneService      7032 non-null   bool   
 8   MultipleLines     7032 non-null   object 
 9   InternetService   7032 non-null   object 
 10  OnlineSecurity    7032 non-null   object 
 11  OnlineBackup      7032 non-null   object 
 12  DeviceProtection  7032 non-null   object 
 13  TechSupport       7032 non-null   object 
 14  StreamingTV       7032 non-null   object 
 15  StreamingMovies   7032 non-null   object 
 16  Contract          7032 non-null   object 
 17  

In [19]:
df_normalizado.sample(6)

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
5389  7359-WWYJV  False    Male          False     True       False      72   
4521  6198-RTPMF  False  Female          False     True       False      17   
3599  4957-TIALW  False  Female          False    False        True      15   
360   0516-WJVXC   True  Female          False    False       False       5   
2986  4139-JPIAM  False    Male          False    False       False      51   
5688  7774-OJSXI   True    Male          False    False       False      31   

      PhoneService     MultipleLines InternetService  ... OnlineBackup  \
5389          True               Yes     Fiber optic  ...          Yes   
4521          True               Yes     Fiber optic  ...           No   
3599          True                No             DSL  ...          Yes   
360           True                No             DSL  ...           No   
2986         False  No phone service             DSL  ...           No   
5688          True               Yes     Fiber optic  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
5389              Yes         Yes         Yes             Yes        Two year   
4521               No          No         Yes             Yes  Month-to-month   
3599              Yes         Yes          No              No        One year   
360                No         Yes          No              No  Month-to-month   
2986               No         Yes          No             Yes  Month-to-month   
5688              Yes          No         Yes             Yes        One year   

     PaperlessBilling            PaymentMethod Charges.Monthly  Charges.Total  
5389            False  Credit card (automatic)          114.45        8375.05  
4521            False         Electronic check           92.60        1579.70  
3599            False  Credit card (automatic)           65.60        1010.00  
360             False         Electronic check           54.20         308.25  
2986             True  Credit card (automatic)           44.45        2181.55  
5688             True         Electronic check          103.45        3066.45  

[6 rows x 21 columns]

### **Tarea 6** Renombrar columnas.

In [24]:
# Para cambiar múltiples columnas
df_normalizado = df_normalizado.rename(columns={
    'customerID': 'id_cliente',
    'gender': 'genero'
    'Churn': 'cliente_actual',
    'SeniorCitizen': 'ciudadano_mayor',
    'Partner': 'pareja',
    'Dependents': 'dependientes',
    'tenure': 'permanencia_meses', # cambiar datos de columna
    'PhoneService': 'servicio_telefono',
    'MultipleLines': 'lineas_multiples',
    'InternetService': 'servicio_internet',
    'OnlineSecurity': 'seguridad_online',
    'OnlineBackup': 'respaldo_online',
    'DeviceProtection': 'proteccion_dispositivo',
    'TechSupport': 'soporte_tecnico',
    'StreamingTV': 'streaming_tv',
    'StreamingMovies': 'streaming_peliculas',
    'Contract': 'contrato',
    'PaperlessBilling': 'facturacion_sin_papel',
    'PaymentMethod': 'metodo_pago',
    'Charges.Monthly': 'cargos_mensuales',
    'Charges.Total': 'cargos_totales'
})

SyntaxError: invalid syntax (ipython-input-24-974463508.py, line 5)

In [21]:
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id_cliente              7032 non-null   object 
 1   cliente_actual          7032 non-null   bool   
 2   gender                  7032 non-null   object 
 3   ciudadano_mayor         7032 non-null   bool   
 4   pareja                  7032 non-null   bool   
 5   dependientes            7032 non-null   bool   
 6   permanencia_meses       7032 non-null   int64  
 7   servicio_telefono       7032 non-null   bool   
 8   lineas_multiples        7032 non-null   object 
 9   servicio_internet       7032 non-null   object 
 10  seguridad_online        7032 non-null   object 
 11  respaldo_online         7032 non-null   object 
 12  proteccion_dispositivo  7032 non-null   object 
 13  soporte_tecnico         7032 non-null   object 
 14  streaming_tv            7032 non-null   objec

### **Tarea 7** Columna de cuentas diarias

Ahora que los datos están limpios, es momento de crear la columna "Cuentas_Diarias". Utiliza la facturación mensual para calcular el valor diario, proporcionando una visión más detallada del comportamiento de los clientes a lo largo del tiempo.



In [22]:
# crear la columna "Cuentas_Diarias" Utilizando la facturación mensual para
# calcular el valor diario

df_normalizado['cargo_por_dia'] = df_normalizado['cargos_mensuales'] / 30
df_normalizado.head()

id_cliente  cliente_actual  gender  ciudadano_mayor  pareja  dependientes  \
0  0002-ORFBO           False  Female            False    True          True   
1  0003-MKNFE           False    Male            False   False         False   
2  0004-TLHLJ            True    Male            False   False         False   
3  0011-IGKFF            True    Male             True    True         False   
4  0013-EXCHZ            True  Female             True    True         False   

   permanencia_meses  servicio_telefono lineas_multiples servicio_internet  \
0                  9               True               No               DSL   
1                  9               True              Yes               DSL   
2                  4               True               No       Fiber optic   
3                 13               True               No       Fiber optic   
4                  3               True               No       Fiber optic   

   ... proteccion_dispositivo soporte_tecnico streaming_tv  \
0  ...                     No             Yes          Yes   
1  ...                     No              No           No   
2  ...                    Yes              No           No   
3  ...                    Yes              No          Yes   
4  ...                     No             Yes          Yes   

  streaming_peliculas        contrato facturacion_sin_papel       metodo_pago  \
0                  No        One year                  True      Mailed check   
1                 Yes  Month-to-month                 False      Mailed check   
2                  No  Month-to-month                  True  Electronic check   
3                 Yes  Month-to-month                  True  Electronic check   
4                  No  Month-to-month                  True      Mailed check   

   cargos_mensuales cargos_totales  cargo_por_dia  
0              65.6         593.30       2.186667  
1              59.9         542.40       1.996667  
2              73.9         280.85       2.463333  
3              98.0        1237.85       3.266667  
4              83.9         267.40       2.796667  

[5 rows x 22 columns]

### **Tarea 8**  Estandarización y transformación de datos.

La estandarización y transformación de datos es una etapa opcional, pero altamente recomendada, ya que busca hacer que la información sea más consistente, comprensible y adecuada para el análisis. Durante esta fase, por ejemplo, puedes convertir valores textuales como "Sí" y "No" en valores binarios (1 y 0), lo que facilita el procesamiento matemático y la aplicación de modelos analíticos.

Además, traducir o renombrar columnas y datos hace que la información sea más accesible y fácil de entender, especialmente cuando se trabaja con fuentes externas o términos técnicos. Aunque no es un paso obligatorio, puede mejorar significativamente la claridad y comunicación de los resultados, facilitando la interpretación y evitando confusiones, especialmente al compartir información con stakeholders no técnicos.

In [23]:
# Ver valores únicos y sus frecuencias para todas las columnas

for col in df_normalizado.columns:
    print(f"--- {col} ---")
    print(df_normalizado[col].value_counts(dropna=False))  # incluye NaN si existen
    print()

--- id_cliente ---
id_cliente
9995-HOTOH    1
0002-ORFBO    1
0003-MKNFE    1
0004-TLHLJ    1
0011-IGKFF    1
             ..
0018-NYROU    1
0017-IUDMW    1
0017-DINOC    1
0016-QLJIS    1
0015-UOCOJ    1
Name: count, Length: 7032, dtype: int64

--- cliente_actual ---
cliente_actual
False    5163
True     1869
Name: count, dtype: int64

--- gender ---
gender
Male      3549
Female    3483
Name: count, dtype: int64

--- ciudadano_mayor ---
ciudadano_mayor
False    5890
True     1142
Name: count, dtype: int64

--- pareja ---
pareja
False    3639
True     3393
Name: count, dtype: int64

--- dependientes ---
dependientes
False    4933
True     2099
Name: count, dtype: int64

--- permanencia_meses ---
permanencia_meses
1     613
72    362
2     238
3     200
4     176
     ... 
38     59
28     57
39     56
44     51
36     50
Name: count, Length: 72, dtype: int64

--- servicio_telefono ---
servicio_telefono
True     6352
False     680
Name: count, dtype: int64

--- lineas_multiples ---
lin

In [25]:
# Cambiar el nombre de la columna gender por género

df_normalizado = df_normalizado.rename(columns={'gender': 'genero'})

In [26]:
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id_cliente              7032 non-null   object 
 1   cliente_actual          7032 non-null   bool   
 2   genero                  7032 non-null   object 
 3   ciudadano_mayor         7032 non-null   bool   
 4   pareja                  7032 non-null   bool   
 5   dependientes            7032 non-null   bool   
 6   permanencia_meses       7032 non-null   int64  
 7   servicio_telefono       7032 non-null   bool   
 8   lineas_multiples        7032 non-null   object 
 9   servicio_internet       7032 non-null   object 
 10  seguridad_online        7032 non-null   object 
 11  respaldo_online         7032 non-null   object 
 12  proteccion_dispositivo  7032 non-null   object 
 13  soporte_tecnico         7032 non-null   object 
 14  streaming_tv            7032 non-null   objec

In [27]:
# Reemplazamos el contenido de las columnas por el idioma español para darle
# más accesibilidad a los socios

# Traducir columna genero
df_normalizado['genero'] = df_normalizado['genero'].replace({
    'Male': 'Masculino',
    'Female': 'Femenino'
})

# Traducir columna lineas_multiples
df_normalizado['lineas_multiples'] = df_normalizado['lineas_multiples'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No phone service': 'Sin servicio telefónico'
})

# Traducir columna servicio_internet
df_normalizado['servicio_internet'] = df_normalizado['servicio_internet'].replace({
    'Fiber optic': 'Fibra óptica',
    'DSL': 'DSL',
    'No': 'No'
})

# Traducir columna seguridad_online
df_normalizado['seguridad_online'] = df_normalizado['seguridad_online'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna respaldo_online
df_normalizado['respaldo_online'] = df_normalizado['respaldo_online'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna proteccion_dispositivo
df_normalizado['proteccion_dispositivo'] = df_normalizado['proteccion_dispositivo'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna soporte_tecnico
df_normalizado['soporte_tecnico'] = df_normalizado['soporte_tecnico'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna streaming_tv
df_normalizado['streaming_tv'] = df_normalizado['streaming_tv'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna streaming_peliculas
df_normalizado['streaming_peliculas'] = df_normalizado['streaming_peliculas'].replace({
    'No': 'No',
    'Yes': 'Sí',
    'No internet service': 'Sin servicio de internet'
})

# Traducir columna contrato
df_normalizado['contrato'] = df_normalizado['contrato'].replace({
    'Month-to-month': 'Mes a mes',
    'Two year': 'Dos años',
    'One year': 'Un año'
})

# Traducir columna metodo_pago
df_normalizado['metodo_pago'] = df_normalizado['metodo_pago'].replace({
    'Electronic check': 'Cheque electrónico',
    'Mailed check': 'Cheque por correo',
    'Bank transfer (automatic)': 'Transferencia bancaria (automática)',
    'Credit card (automatic)': 'Tarjeta de crédito (automática)'
})

In [28]:
df_normalizado.head(10)

id_cliente  cliente_actual     genero  ciudadano_mayor  pareja  \
0  0002-ORFBO           False   Femenino            False    True   
1  0003-MKNFE           False  Masculino            False   False   
2  0004-TLHLJ            True  Masculino            False   False   
3  0011-IGKFF            True  Masculino             True    True   
4  0013-EXCHZ            True   Femenino             True    True   
5  0013-MHZWF           False   Femenino            False   False   
6  0013-SMEOE           False   Femenino             True    True   
7  0014-BMAQU           False  Masculino            False    True   
8  0015-UOCOJ           False   Femenino             True   False   
9  0016-QLJIS           False   Femenino            False    True   

   dependientes  permanencia_meses  servicio_telefono lineas_multiples  \
0          True                  9               True               No   
1         False                  9               True               Sí   
2         False                  4               True               No   
3         False                 13               True               No   
4         False                  3               True               No   
5          True                  9               True               No   
6         False                 71               True               No   
7         False                 63               True               Sí   
8         False                  7               True               No   
9          True                 65               True               Sí   

  servicio_internet  ... proteccion_dispositivo soporte_tecnico streaming_tv  \
0               DSL  ...                     No              Sí           Sí   
1               DSL  ...                     No              No           No   
2      Fibra óptica  ...                     Sí              No           No   
3      Fibra óptica  ...                     Sí              No           Sí   
4      Fibra óptica  ...                     No              Sí           Sí   
5               DSL  ...                     No              Sí           Sí   
6      Fibra óptica  ...                     Sí              Sí           Sí   
7      Fibra óptica  ...                     No              Sí           No   
8               DSL  ...                     No              No           No   
9               DSL  ...                     Sí              Sí           Sí   

  streaming_peliculas   contrato facturacion_sin_papel  \
0                  No     Un año                  True   
1                  Sí  Mes a mes                 False   
2                  No  Mes a mes                  True   
3                  Sí  Mes a mes                  True   
4                  No  Mes a mes                  True   
5                  Sí  Mes a mes                  True   
6                  Sí   Dos años                  True   
7                  No   Dos años                  True   
8                  No  Mes a mes                  True   
9                  Sí   Dos años                  True   

                           metodo_pago  cargos_mensuales cargos_totales  \
0                    Cheque por correo             65.60         593.30   
1                    Cheque por correo             59.90         542.40   
2                   Cheque electrónico             73.90         280.85   
3                   Cheque electrónico             98.00        1237.85   
4                    Cheque por correo             83.90         267.40   
5      Tarjeta de crédito (automática)             69.40         571.45   
6  Transferencia bancaria (automática)            109.70        7904.25   
7      Tarjeta de crédito (automática)             84.65        5377.80   
8                   Cheque electrónico             48.20         340.35   
9                    Cheque por correo             90.45        5957.90   

   cargo_por_dia  
0       2.186667  
1       1.996667  
2       2.463333  
3     

##📊 Carga y análisis

##📄Informe final